README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# scBSC (single cell Bivariate Spatial Correlation)

<!-- badges: start -->
<!-- badges: end -->

scBSC allows users calculate bivariate spatial correlation statistics as
applied from [Lee S., J Geography Syst
(2001)](https://link.springer.com/article/10.1007/s101090100064) to this
wonderful spatial single cell RNA-Seq domain.

## Installation

You can install the released version of scBSC from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("scBSC")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("soymintc/scBSC")
```

## Example

The following calculates spatial correlation between expression of two
genes, Penk and Cck, from a Seurat-format mouse brain spatial scRNA-seq
data from Visium 10x.

#### Using pre-made SeuratData

``` r
# Import libraries
library(Seurat)
library(SeuratData)
library(scBSC)

# Import Seurat-format mouse brain data
InstallData("stxBrain")
brain <- LoadData("stxBrain", type = "anterior1")
brain <- SCTransform(brain, assay = "Spatial", verbose = FALSE)

# Calculate bivariate spatial correlation statistics
conn_mat <- make_conn_mat(brain) # creating connectivity matrix; only required once per dataset
bsc <- calc_bsc("Penk", "Cck", brain, conn_mat) # calculate statistics for any pair of genes in the data
print_ <- sprintf("%.3f %.3f %.3f %.3f %.3f",
  bsc$L_XX, # How well is Penk expression spatially clustered
  bsc$L_YY, # How well is Cck expression spatially clustered
  bsc$r_sm, # Correlation between spatially smoothened expression of two genes
  bsc$r, # Correlation between expression of two genes
  bsc$L_XY) # Bivariate spatial correlation (clustering effect & expression correlation)
print(print_) # result: 0.869 0.865 -0.809 -0.729 -0.702
```

#### Using Visium 10x data, imported by Seurat Load10X\_Spatial

Of course scBSC also supports your own spatial scRNA-seq, as long as the
data is in Seurat spatial object format. Link to raw data: [Visium 10x
Mouse Brain Serial
Section](https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Mouse_Brain_Sagittal_Anterior)

``` r
# Import libraries
library(Seurat)
library(scBSC)

# Import Visium 10x data using Seurat
setwd('/path/to/mouse_brain_data')
# Use Load10X_Spatial to load data from the following file structure.
# Data used was Visium 10x Mouse Brain Serial Section 1 (Anterior-Sagittal)
# anterior1 (directory under /path/to/mouse_brain_data; any name would suffice)
# ├── filtered_feature_bc_matrix
# │   ├── barcodes.tsv.gz # cell barcodes
# │   ├── features.tsv.gz # gene info
# │   └── matrix.mtx.gz # gene count per cell
# ├── filtered_feature_bc_matrix.h5 # [req] sparse matrix including the info above
# └── spatial
#     ├── aligned_fiducials.jpg
#     ├── detected_tissue_image.jpg
#     ├── scalefactors_json.json # [req] high -> low-res image scaling info
#     ├── tissue_hires_image.png
#     ├── tissue_lowres_image.png # [req] 
#     └── tissue_positions_list.csv # [req] the coordinates for each spot
brain <- Load10X_Spatial(data.dir="./anterior1",
                         filename="filtered_feature_bc_matrix.h5",
                         assay='Spatial',
                         slice="anterior1",
                         filter.matrix = TRUE)
brain@meta.data$orig.ident <- "anterior1" # give cells identity; default: SeuratProject
Idents(object = brain) <- "anterior1" # label experiment
Project(object = brain) <- "brain" # label project

# Remove some unwanted genes for your downstream experiments
brain <- PercentageFeatureSet(brain, "^mt-", col.name = "percent_mito")
brain <- PercentageFeatureSet(brain, "^Hb.*-", col.name = "percent_hb")
brain = brain[, brain$nFeature_Spatial > 500 &
                brain$percent_mito < 25 &
                brain$percent_hb < 20]
brain <- brain[!grepl("Bc1", rownames(brain)), ] # filter Bc1
brain <- brain[!grepl("^mt-", rownames(brain)), ] # filter Mitocondrial
brain <- brain[!grepl("^Hb.*-", rownames(brain)), ] # filter Hemoglobin gene (optional if Hb genes needed)

# Normalize expression values by SCTransform
# Required for later calculations based on assay="SCT"; uses brain@assays[["SCT"]]@data as default
# Not required if the assay property in calc_bsc is to be changed to "Spatial" (then only requires brain@assays[["Spatial"]])
brain <- SCTransform(brain, assay = "Spatial", verbose = TRUE, method = "poisson")

# Calculate bivariate spatial correlation statistics
conn_mat <- make_conn_mat(brain) # creating connectivity matrix; only required once per dataset
bsc <- calc_bsc("Penk", "Cck", brain, conn_mat) # calculate statistics for any pair of genes in the data
print_ <- sprintf("%.3f %.3f %.3f %.3f %.3f",
  bsc$L_XX, # How well is Penk expression spatially clustered
  bsc$L_YY, # How well is Cck expression spatially clustered
  bsc$r_sm, # Correlation between spatially smoothened expression of two genes
  bsc$r, # Correlation between expression of two genes
  bsc$L_XY) # Bivariate spatial correlation (clustering effect & expression correlation)
print(print_) # result: 0.869 0.865 -0.809 -0.729 -0.702
```

Plotting the relative gene expression between the two genes will show
the following result. Below includes a highly negative, highly positive,
and close-to-zero Lxy cases.

``` r
# Highly positive spatial correlation example with Lxx 0.869, Lyy 0.864, rsm 0.936, r 0.855, Lxy 0.813
SpatialFeaturePlot(brain, features = c("Gpr88", "Ppp1r1b"), ncol = 2, alpha = c(0.1, 1), max.cutoff = 5)
```

![Expression between two
genes](man/figures/README-1.png?raw=true "Highly positive spatial correlation")

``` r
# Highly negative spatial correlation with Lxx 0.869, Lyy 0.865, rsm -0.809, r -0.729, Lxy -0.702
SpatialFeaturePlot(brain, features = c("Penk", "Cck"), ncol = 2, alpha = c(0.1, 1), max.cutoff = 5)
```

![Expression between two
genes](man/figures/README-2.png?raw=true "Highly negative spatial correlation")

``` r
# Close-to-zero spatial correlation with Lxx 0.461, Lyy 0.569, rsm 0.075, r 0.059, Lxy 0.038
SpatialFeaturePlot(brain, features = c("Rpl34", "Ptn"), ncol = 2, alpha = c(0.1, 1), max.cutoff = 5)
```

![Expression between two
genes](man/figures/README-3.png?raw=true "Close-to-zero spatial correlation")


NEWS.md
# scBSC 0.1.0

* Initial build

# scBSC 0.1.1

* Added a `NEWS.md` file to track changes to the package.

# scBSC 0.1.2

* Fixed examples to refer to github page and readme to include more figures.

# scBSC 0.1.3

* Added class slot and list validations.


R/make_conn_mat.R
#' Returns a list of spatial connectivity matrix variables (C, W)
#' according to Lee S. J Geograph Syst (2001) 3:369-385.
#'
#' @title Make connectivity matrix
#' @name make_conn_mat
#' @description This function calculates the connectivity of a spatial cell to other cells.
#' @param sdata Seurat formatted spatial 10X scRNA expression data
#' @param orig.ident A unique orig.ident for creating connectivity matrix.
#' @param verbose Print intermediate value checks as stderr()
#' @return A list containing $barcodes_in_tissue (barcodes of spatial cells with 'tissue=TRUE'),
#' $nbarcodes_in_tissue (length of barcodes_in_tissue),
#' $C (raw connectivity/adjacency matrix),
#' $W (row-sum divided connectivity matrix),
#' $L_estimate_divR (L estimate, divided by Pearson's R (used later))
#' @import Matrix
#' @import dplyr
#' @import methods
#' @export
#' @examples #' # Refer to https://github.com/soymintc/scBSC

make_conn_mat = function(sdata, orig.ident=NULL, verbose=F) { # input Seurat spatial data
  # Seurat - sdata@images$anterior1@coordinates format:
  #                    tissue row col imagerow imagecol
  # AAACAAGTATCTCCCA-1      1  50 102     7474     8500
  # AAACACCAATAACTGC-1      1  59  19     8552     2788
  # AAACAGAGCGACTCCT-1      1  14  94     3163     7950
  # AAACAGCTTTCAGAAG-1      1  43   9     6636     2100
  if (!.hasSlot(sdata, "meta.data")) stop("[ERROR] data class does not contain @meta.data")
  if (is.null(sdata@meta.data$orig.ident)) stop(paste("[ERROR] @meta.data do not contain $orig.ident"))
  if (length(unique(sdata@meta.data$orig.ident)) > 1) stop("[ERROR] $orig.ident is not unique.")
  if (is.null(orig.ident)) orig.ident = unique(sdata@meta.data$orig.ident)[1]
  if (!.hasSlot(sdata@images[[orig.ident]], "coordinates")) stop("[ERROR] data class does not contain @coordinates")
  if (dim(sdata@images[[orig.ident]]@coordinates)[1] == 0) stop("[ERROR] @coordinates count is zero.")
  coordinates = sdata@images[[orig.ident]]@coordinates
  positions_in_tissue = coordinates %>% filter(tissue==1)
  if (verbose) print(sprintf('dim(positions_in_tissue): %s', dim(positions_in_tissue)))
  barcodes_in_tissue = rownames(positions_in_tissue)
  if (verbose) print(sprintf('length(barcodes_in_tissue): %d', length(barcodes_in_tissue)))
  nbarcodes_in_tissue = length(barcodes_in_tissue)
  if (verbose) print(sprintf('nbarcodes_in_tissue: %d', nbarcodes_in_tissue))

  C = Matrix(nrow=nbarcodes_in_tissue, ncol=nbarcodes_in_tissue, data=0, sparse=T) # C: (sparse) connectivity matrix
  if (!exists("C")) stop("[ERROR] inner matrix C does not exist")
  if (dim(C)[1] == 0) stop("[ERROR] row length of C == 0")
  if (verbose) print(sprintf('dim(C): %s', dim(C)))
  rownames(C) = barcodes_in_tissue
  colnames(C) = barcodes_in_tissue

  for (barcode in barcodes_in_tissue) {
    conn_mat = list() # data holder for connectivity matrices
    row_i = positions_in_tissue[barcode, 'row']
    col_i = positions_in_tissue[barcode, 'col'] # sdata$counts[['A']][[barcode]] later
    # Set nearby nodes and check connectivity (only in_tissue)
    neighbors = subset(coordinates,
                       tissue==1 &
                         (
                           ((row==row_i-1) & (col==col_i-1)) |
                             ((row==row_i-1) & (col==col_i+1)) |
                             ((row==row_i) & (col==col_i-2)) |
                             ((row==row_i) & (col==col_i+2)) |
                             ((row==row_i+1) & (col==col_i-1)) |
                             ((row==row_i+1) & (col==col_i+1))
                         ))
    neighbor_barcodes = rownames(neighbors)
    if (length(neighbor_barcodes) > 0) C[barcode, neighbor_barcodes] = 1
  }
  if (verbose) print(sprintf('dim(C): %s', dim(C)))
  if (verbose) print(sprintf('head(summary(C)): %s', head(summary(C))))

  W = C / rowSums(C) # W: weighted connectivity matrix
  if (!exists("W")) stop("[ERROR] inner matrix W does not exist")
  if (dim(W)[1] == 0) stop("[ERROR] row length of W == 0")
  if (verbose) print(sprintf('head(summary(W)): %s', head(summary(W))))
  W[is.na(W)] = 0

  conn_mat[['L_estimate_divR']] = (sum(diag(t(W %*% W)))
                                   / (nbarcodes_in_tissue - 1))

  conn_mat[['barcodes_in_tissue']] = barcodes_in_tissue
  conn_mat[['nbarcodes_in_tissue']] = nbarcodes_in_tissue
  conn_mat[['W']] = W
  conn_mat[['C']] = C

  return(conn_mat)
}


R/calc_bsc.R
#' Returns a list of bivariate spatial correlation statistics
#' according to Lee S. J Geograph Syst (2001) 3:369-385.
#'
#' @title Calculate Bivariate Spatial Correlation
#' @name calc_bsc
#' @description This function converts input temperatures in Fahrenheit to Celsius.
#' @param feature1 a gene symbol as in Seurat object
#' @param feature2 another gene symbol as in Seurat object
#' @param sdata Seurat formatted spatial 10X scRNA expression data
#' @param conn_mat connectivity matrix (conn_mat) from make_conn_mat
#' @param assay default to 'SCT' but can be 'Spatial' (raw) or 'SCT'
#' @return A list containing $r (Pearson's r),
#' $r_sm (spatially smoothened r),
#' $L_XX (clustering factor of feature1),
#' $L_YY (clustering factor of feature2),
#' $L_XY (L_X,Y bivariate spatial correlation from)
#' @import Matrix
#' @import dplyr
#' @import methods
#' @export
#' @examples # Refer to https://github.com/soymintc/scBSC
calc_bsc = function(feature1, feature2, sdata, conn_mat, assay='SCT') {
  # Set X, Y variables from two features
  # Seurat - sdata@assays$SCT@data: row=feature, col=cell matrix
  # feature1 = 'Gpr88' # debug
  # feature2 = 'Penk' # debug
  if (!.hasSlot(sdata, "assays")) stop("[ERROR] data class does not contain @assays")
  if (is.null(sdata@assays[[assay]])) stop(paste("[ERROR] @assays do not contain", assay))
  if (!.hasSlot(sdata@assays[[assay]], "data")) stop("[ERROR] data class does not contain @data")
  mrna = sdata@assays[[assay]]@data # default to assay='SCT'
  X_values = mrna[feature1, conn_mat$barcodes_in_tissue] # 1:X
  Y_values = mrna[feature2, conn_mat$barcodes_in_tissue] # 2:Y
  X = data.frame(value=X_values)
  Xmean = mean(X$value)
  Y = data.frame(value=Y_values)
  Ymean = mean(Y$value)

  # Smoothened values
  X['smooth'] = conn_mat$W %*% X[,'value'] # Xsm = W * X
  Y['smooth'] = conn_mat$W %*% Y[,'value'] # Ysm = W * Y
  Xmean_sm = mean(X$smooth) # muX
  Ymean_sm = mean(Y$smooth) # muY

  # Calculate Peason's r(X,Y), r(smooth), L_XX, L_YY, L_XY as in Lee S (2001)
  r = (sum((X$value - Xmean) * (Y$value - Ymean))
       / (sqrt(sum((X$value - Xmean)^2)) * sqrt(sum((Y$value - Ymean)^2))))
  r_sm = (sum((X$smooth - Xmean_sm) * (Y$smooth - Ymean_sm))
          / (sqrt(sum((X$smooth - Xmean_sm)^2)) * sqrt(sum((Y$smooth - Ymean_sm)^2))))
  L_XX = (sum((X$smooth - Xmean)^2) / sum((X$value - Xmean)^2))
  L_YY = (sum((Y$smooth - Ymean)^2) / sum((Y$value - Ymean)^2))
  L_XY = sqrt(L_XX) * sqrt(L_YY) * r_sm

  # Group into Bivariate Spatial Correlation values list
  bsc = list('r' = r, 'r_sm' = r_sm,
             'L_XX' = L_XX, 'L_YY' = L_YY, 'L_XY' = L_XY)

  return(bsc)
}
